packages/@mantine/modals/src/ModalsProvider.tsx

```diff
@@ -112,55 +103,72 @@ export function ModalsProvider({ children, modalProps, labels, modals }: ModalsP
     [dispatch]
   );
 
-  const openContextModal = useCallback(
-    (modal: string, { modalId, ...props }: OpenContextModal) => {
+  const openContextModal: ModalsContextProps['openContextModal'] = useCallback(
+    ({ modalId, modalKey, ...props }) => {
       const id = modalId || randomId();
       dispatch({
         type: 'OPEN',
         modal: {
           id,
           type: 'context',
           props,
-          ctx: modal,
+          ctx: modalKey,
         },
       });
       return id;
     },
     [dispatch]
   );
 
-  const closeModal = useCallback(
-    (id: string, canceled?: boolean) => {
+  const closeModal: ModalsContextProps['closeModal'] = useCallback(
+    (modalId: string, canceled?: boolean) => {
+      dispatch({ type: 'CLOSE', modalId, canceled });
+    },
+    [stateRef, dispatch]
+  );
+
+  const closeContextModal: ModalsContextProps['closeContextModal'] = useCallback(
+    (modalKey: string, canceled?: boolean) => {
+      const id =
+        stateRef.current.modals.find((m) => m.type === 'context' && m.ctx === modalKey)?.id ??
+        modalKey;
       dispatch({ type: 'CLOSE', modalId: id, canceled });
     },
     [stateRef, dispatch]
   );
 
-  const updateModal = useCallback(
-    ({ modalId, ...newProps }: Partial<ModalSettings> & { modalId: string }) => {
-      dispatch({
-        type: 'UPDATE',
-        modalId,
-        newProps,
-      });
+  const closeAllModals: ModalsContextProps['closeAllModals'] = useCallback(
+    (canceled?: boolean) => {
+      dispatch({ type: 'CLOSE_ALL', canceled });
     },
-    [dispatch]
+    [stateRef, dispatch]
   );
 
-  const updateContextModal = useCallback(
-    ({ modalId, ...newProps }: { modalId: string } & Partial<OpenContextModal<any>>) => {
+  const updateModal: ModalsContextProps['updateModal'] = useCallback(
+    ({ modalId, ...newProps }) => {
       dispatch({ type: 'UPDATE', modalId, newProps });
     },
     [dispatch]
   );
 
+  const updateContextModal: ModalsContextProps['updateContextModal'] = useCallback(
+    ({ modalKey, modalId, ...newProps }) => {
+      const id =
+        modalId ??
+        stateRef.current.modals.find((m) => m.type === 'context' && m.ctx === modalKey)?.id ??
+        modalKey;
```

Comment:
I'm unsure if you want to have that kind of logic inside the Provider of if you prefer to have it inside the reducer. But doing it in the reducer may require a new dispatch action for updateContext like 'UPDATE_CONTEXT'